### PR TITLE
feat(trie): witness

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Ref https://github.com/paradigmxyz/reth/pull/9249

Create a trie witness implementation that gathers all trie nodes needed to apply state transition.

### CODE REVIEW

File: crates/trie/trie/src/witness.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            let storage_multiproof = account_multiproof
+                .storages
+                .get(&hashed_address)
+                .ok_or(TrieWitnessError::MissingStorageMultiProof(hashed_address))?;
+
+            // Gather and record account trie nodes.
+            let account = state
+                .accounts
+                .get(&hashed_address)
+                .ok_or(TrieWitnessError::MissingAccount(hashed_address))?;
```
Comment (mattsse) [inline crates/trie/trie/src/witness.rs:98]: or_else?

File: crates/trie/trie/src/witness.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            let value = if account.is_some() || storage_multiproof.root != EMPTY_ROOT_HASH {
+                account_rlp.clear();
+                TrieAccount::from((account.unwrap_or_default(), storage_multiproof.root))
+                    .encode(&mut account_rlp as &mut dyn BufMut);
```
Comment (mattsse) [inline crates/trie/trie/src/witness.rs:102]: why is this cast required?

**Review by mattsse — APPROVED** [review]
tbh I don't have sufficient context to review this in detail but, logic wise this looks okay to me

File: crates/trie/trie/src/witness.rs (hunk 2 above)
Comment (rkrasiuk) [reply crates/trie/trie/src/witness.rs:102]: to make rust analyzer happy :/

File: crates/trie/trie/src/witness.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/trie/trie/src/witness.rs:98]: just a copy here, is it needed?

**Pull request merged**